Enable strict deserialization of DDActionContext behind a runtime flag

achristensen07 opened the pull request:
#### 45a17f7bba614c5df5b50500bc5f16dcce1215ab
<pre>
Enable strict deserialization of DDActionContext behind a runtime flag
<a href="https://bugs.webkit.org/show_bug.cgi?id=256279">https://bugs.webkit.org/show_bug.cgi?id=256279</a>
rdar://108866400

Reviewed by David Kilzer.

WebKit gets them from 3 places:

1. We allocate them ourselves.  I replaced this with allocation of DDSecureActionContexts.
2. We get them from calling contextForView:altMode:interactionStartedHandler:interactionChangedHandler:interactionStoppedHandler:
   on an existing DDActionContext.  This will give us a DDSecureActionContexts once rdar://108865826 is fixed.
3. We get them from WebUIDelegatePrivate&apos;s _webView:actionContextForHitTestResult:range: in WebKitLegacy.
   This is inconsequential because these will never be serialized across IPC, and DDActionContext and DDSecureActionContext
   behave the same except for their serialization code.  I added a cast in WebKitLegacy to keep it compiling, but ObjC will
   just do the right thing with all selectors given to the object.  Also, there&apos;s only one user of this delegate callback
   and it seems to be dead code.

I verified that adopting DDSecureActionContext causes the tests that used to have issues deserializing a DDActionContext
to no longer have the issues.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/mac/DataDetectorsSoftLink.h:
* Source/WebCore/PAL/pal/mac/DataDetectorsSoftLink.mm:
* Source/WebCore/PAL/pal/spi/mac/DataDetectorsSPI.h:
* Source/WebCore/editing/cocoa/DataDetection.h:
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::detectItem):
* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::menuItemForTelephoneNumber):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::haveSecureActionContext):
(IPC::shouldEnableStrictMode):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/mac/WebHitTestResultPlatformData.serialization.in:
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
* Source/WebKit/UIProcess/mac/WKImmediateActionController.h:
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController _animationControllerForDataDetectedText]):
(-[WKImmediateActionController _animationControllerForDataDetectedLink]):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePageOverlay.cpp:
* Source/WebKit/WebProcess/WebPage/WebPageOverlay.h:
* Source/WebKitLegacy/mac/WebView/WebImmediateActionController.mm:
(-[WebImmediateActionController _animationControllerForDataDetectedText]):

Canonical link: <a href="https://commits.webkit.org/263721@main">https://commits.webkit.org/263721@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/204cdbe49400689f7396642c843b03b888da31cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5503 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5649 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5839 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7048 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5512 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5874 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5623 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6808 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5604 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5664 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4923 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7081 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3141 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4938 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12078 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/4567 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5007 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5018 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6763 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5088 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5454 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4472 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/5609 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4910 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1385 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9013 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5770 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/635 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5272 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1503 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -562,6 +562,17 @@ static void encodeSecureCodingInternal(Encoder& encoder, id <NSObject, NSSecureC
 }
 
+#if ENABLE(DATA_DETECTION) || ENABLE(REVEAL)
+static bool haveSecureActionContext()
```
Why not just make this `inline` instead of `static`?  Seems very inline-able.

achristensen07 replied:
It will be inlined by the optimizer.  "inline" doesn't really do anything in cases like this.

ddkilzer commented on Source/WebCore/editing/cocoa/DataDetection.h:
```diff
@@ -33,7 +33,11 @@
+OBJC_CLASS DDSecureActionContext;
+#else
 OBJC_CLASS DDActionContext;
+#endif
```
You should add something like this to `DataDetectorsSPI.h`:

```
#if HAVE(SECURE_ACTION_CONTEXT)
using WKDDActionContext = DDSecureActionContext;
#else
using WKDDActionContext = DDActionContext;
#endif
```

Then instead of doing the `#if HAVE(SECURE_ACTION_CONTEXT)` dance everywhere, you can just use `WKDDActionConext` instead.  (I'm hoping this works for Obj-C types in both C++ and Obj-C++ contexts, but I've never tried.)

achristensen07 replied:
That works for all the ObjC.  There's only one use in a .cpp file (WKBundlePageOverlay.cpp) which is going to be removed after rdar://108925315 is resolved.

achristensen07 replied:
It also doesn't work for headers, but it cleaned up all the .mm files nicely.

ddkilzer commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -113,7 +113,12 @@
+    auto actionContext = adoptNS([PAL::allocDDSecureActionContextInstance() init]);
+#else
     auto actionContext = adoptNS([PAL::allocDDActionContextInstance() init]);
+#endif
```
You could do something similar (write an inline method) in `DataDetectorsSPI.h` to eliminate this code as well:

```
    inline WKDDActionContext PAL::allocWKDDActionContextInstance()
    {
#if HAVE(SECURE_ACTION_CONTEXT)
        return PAL::allocDDSecureActionContextInstance();
#else
        return PAL::allocDDActionContextInstance();
#endif
    }
```

achristensen07 replied:
I like it.

ddkilzer reviewed (approved):
r=me, but please consider feedback comments.

(None of the changes are required to land, but I think they will increase readability.)

achristensen07 committed:
Enable strict deserialization of DDActionContext behind a runtime flag
https://bugs.webkit.org/show_bug.cgi?id=256279
rdar://108866400

Reviewed by David Kilzer.

WebKit gets them from 3 places:

1. We allocate them ourselves.  I replaced this with allocation of DDSecureActionContexts.
2. We get them from calling contextForView:altMode:interactionStartedHandler:interactionChangedHandler:interactionStoppedHandler:
   on an existing DDActionContext.  This will give us a DDSecureActionContexts once rdar://108865826 is fixed.
3. We get them from WebUIDelegatePrivate's _webView:actionContextForHitTestResult:range: in WebKitLegacy.
   This is inconsequential because these will never be serialized across IPC, and DDActionContext and DDSecureActionContext
   behave the same except for their serialization code.  I added a cast in WebKitLegacy to keep it compiling, but ObjC will
   just do the right thing with all selectors given to the object.  Also, there's only one user of this delegate callback
   and it seems to be dead code.

I verified that adopting DDSecureActionContext causes the tests that used to have issues deserializing a DDActionContext
to no longer have the issues.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/PAL/pal/mac/DataDetectorsSoftLink.h:
* Source/WebCore/PAL/pal/mac/DataDetectorsSoftLink.mm:
* Source/WebCore/PAL/pal/spi/mac/DataDetectorsSPI.h:
* Source/WebCore/editing/cocoa/DataDetection.h:
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::detectItem):
* Source/WebKit/Platform/mac/MenuUtilities.mm:
(WebKit::menuItemForTelephoneNumber):
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::haveSecureActionContext):
(IPC::shouldEnableStrictMode):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/mac/WebHitTestResultPlatformData.serialization.in:
* Source/WebKit/UIProcess/API/APIHitTestResult.h:
* Source/WebKit/UIProcess/mac/WKImmediateActionController.h:
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController _animationControllerForDataDetectedText]):
(-[WKImmediateActionController _animationControllerForDataDetectedLink]):
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePageOverlay.cpp:
* Source/WebKit/WebProcess/WebPage/WebPageOverlay.h:
* Source/WebKitLegacy/mac/WebView/WebImmediateActionController.mm:
(-[WebImmediateActionController _animationControllerForDataDetectedText]):

Canonical link: https://commits.webkit.org/263721@main
